### PR TITLE
Turn on aggressive fusion heuristics by default.

### PR DESCRIPTION
**Pull request by MaheshRavishankar**

### CODE REVIEW

**Comment by MaheshRavishankar** [comment]
Current blocker is #12019

**Comment by MaheshRavishankar** [comment]
All Vulkan and CPU tests pass for me locally now. Running CI.

**Comment by ThomasRaoux** [comment]
I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.

**Comment by MaheshRavishankar** [comment]
> I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.

Didnt you plan to move all of that into the softmax decomposition pass?

**Comment by MaheshRavishankar** [comment]
Going to look into the failures, but as it stands these are the dispatches on the benchmarks 

**Before**

| Model | Fills | Copies | Dispatches | Executables | Generic Before | Generic After | DispatchWorkgroups |
| --- | --- | --- | --- | --- | --- | --- | --- |
| bert_encoderdefault_off | 11 | 4 | 514 | 41 | 1198 | 534 | 514 |
| collatzdefault_off | 2 | 1 | 0 | 41 | 12 | 0 | 0 |
| deeplabdefault_off | 18 | 1 | 78 | 41 | 105 | 57 | 78 |
| edge_detectiondefault_off | 1 | 0 | 3 | 41 | 4 | 1 | 3 |
| fragmentdefault_off | 3 | 1 | 1 | 41 | 10 | 1 | 1 |
| fullyconnecteddefault_off | 0 | 0 | 3 | 41 | 8 | 3 | 3 |
| minilmdefault_off | 0 | 2 | 223 | 41 | 1137 | 249 | 223 |
| mnistdefault_off | 4 | 0 | 6 | 41 | 10 | 6 | 6 |
| mobilebert-fp32default_off | 25 | 4 | 752 | 41 | 1741 | 652 | 752 |
| mobilenetv2default_off | 18 | 1 | 74 | 41 | 103 | 56 | 74 |
| mobilessddefault_off | 30 | 1 | 121 | 41 | 157 | 88 | 121 |
| persondetect-int8default_off | 14 | 1 | 75 | 41 | 95 | 45 | 75 |
| posenetdefault_off | 14 | 1 | 48 | 41 | 59 | 32 | 48 |
| resnet50default_off | 338 | 0 | 78 | 41 | 711 | 55 | 78 |
| unidirectional_lstmdefault_off | 5 | 5 | 10 | 41 | 43 | 7 | 10 |

**After**
| Model | Fills | Copies | Dispatches | Executables | Generic Before | Generic After | DispatchWorkgroups |
| --- | --- | --- | --- | --- | --- | --- | --- |
| bert_encoderdefault_on | 11 | 4 | 489 | 41 | 1198 | 533 | 489 |
| collatzdefault_on | 2 | 1 | 0 | 41 | 12 | 0 | 0 |
| deeplabdefault_on | 18 | 1 | 78 | 41 | 105 | 57 | 78 |
| edge_detectiondefault_on | 1 | 0 | 3 | 41 | 4 | 1 | 3 |
| fragmentdefault_on | 3 | 1 | 1 | 41 | 10 | 1 | 1 |
| fullyconnecteddefault_on | 0 | 0 | 3 | 41 | 8 | 3 | 3 |
| minilmdefault_on | 0 | 2 | 184 | 41 | 1137 | 248 | 184 |
| mnistdefault_on | 4 | 0 | 4 | 41 | 10 | 6 | 4 |
| mobilebert-fp32default_on | 1 | 4 | 727 | 41 | 1741 | 651 | 727 |
| mobilenetv2default_on | 18 | 1 | 73 | 41 | 103 | 56 | 73 |
| mobilessddefault_on | 30 | 1 | 121 | 41 | 157 | 88 | 121 |
| persondetect-int8default_on | 14 | 1 | 58 | 41 | 95 | 45 | 58 |
| posenetdefault_on | 14 | 1 | 48 | 41 | 59 | 32 | 48 |
| resnet50default_on | 338 | 0 | 76 | 41 | 711 | 55 | 76 |
| unidirectional_lstmdefault_on | 5 | 8 | 8 | 41 | 43 | 5 | 8 |

**Comment by ThomasRaoux** [comment]
> > I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.
> 
> Didnt you plan to move all of that into the softmax decomposition pass?

Right, the idea is to make the softmax decomposition do the rematerialization when decomposing and then remove the general pass. That means if a code sequence like softmax (but not using the softmax op) is fused we will have large allocations again. It sounds like that’s what could happen with aggressive fusion.

**Comment by MaheshRavishankar** [comment]
> > > I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.
> > 
> > 
> > Didnt you plan to move all of that into the softmax decomposition pass?
> 
> Right, the idea is to make the softmax decomposition do the rematerialization when decomposing and then remove the general pass. That means if a code sequence like softmax (but not using the softmax op) is fused we will have large allocations again. It sounds like that’s what could happen with aggressive fusion.

yup, understood... today rematerialization works, cause really the only case that get fused that way is softmax. Once the decomposition does rematerialzation, I can remove the use of that pass in the backends and make sure we dont fuse in a way that creates allocations. I think those could be decoupled. In any case, large stacks are a hard-error today... so we will know if there are issues.

**Comment by ThomasRaoux** [comment]
> > > > I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.
> > > 
> > > 
> > > Didnt you plan to move all of that into the softmax decomposition pass?
> > 
> > 
> > Right, the idea is to make the softmax decomposition do the rematerialization when decomposing and then remove the general pass. That means if a code sequence like softmax (but not using the softmax op) is fused we will have large allocations again. It sounds like that’s what could happen with aggressive fusion.
> 
> yup, understood... today rematerialization works, cause really the only case that get fused that way is softmax. Once the decomposition does rematerialzation, I can remove the use of that pass in the backends and make sure we dont fuse in a way that creates allocations. I think those could be decoupled. In any case, large stacks are a hard-error today... so we will know if there are issues.

Got it, so you are going to remove the cases that would create large allocs from the fusion? That would work. This is not done yet right?

**Comment by MaheshRavishankar** [comment]
> > > > > I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.
> > > > 
> > > > 
> > > > Didnt you plan to move all of that into the softmax decomposition pass?
> > > 
> > > 
> > > Right, the idea is to make the softmax decomposition do the rematerialization when decomposing and then remove the general pass. That means if a code sequence like softmax (but not using the softmax op) is fused we will have large allocations again. It sounds like that’s what could happen with aggressive fusion.
> > 
> > 
> > yup, understood... today rematerialization works, cause really the only case that get fused that way is softmax. Once the decomposition does rematerialzation, I can remove the use of that pass in the backends and make sure we dont fuse in a way that creates allocations. I think those could be decoupled. In any case, large stacks are a hard-error today... so we will know if there are issues.
> 
> Got it, so you are going to remove the cases that would create large allocs from the fusion? That would work. This is not done yet right?

Trying to guage the order of things. Ill probably defer that to later stage. So keep the rematerialization for now (its used on all backends), and then evaluate after softmax doesnt need it.

**Comment by ThomasRaoux** [comment]
> > > > > > I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.
> > > > > 
> > > > > 
> > > > > Didnt you plan to move all of that into the softmax decomposition pass?
> > > > 
> > > > 
> > > > Right, the idea is to make the softmax decomposition do the rematerialization when decomposing and then remove the general pass. That means if a code sequence like softmax (but not using the softmax op) is fused we will have large allocations again. It sounds like that’s what could happen with aggressive fusion.
> > > 
> > > 
> > > yup, understood... today rematerialization works, cause really the only case that get fused that way is softmax. Once the decomposition does rematerialzation, I can remove the use of that pass in the backends and make sure we dont fuse in a way that creates allocations. I think those could be decoupled. In any case, large stacks are a hard-error today... so we will know if there are issues.
> > 
> > 
> > Got it, so you are going to remove the cases that would create large allocs from the fusion? That would work. This is not done yet right?
> 
> Trying to guage the order of things. Ill probably defer that to later stage. So keep the rematerialization for now (its used on all backends), and then evaluate after softmax doesnt need it.

Sounds good to me. I was going to remove it because you were concerned about it but I can’t really think about case where doing the rematerialization is worse and where backend cse wouldn’t revert it in case it is worth.
I guess we could have saved the work in matching softmax if we are going to enable aggressive fusion anyway

**Comment by MaheshRavishankar** [comment]
> > > > > > > I thought you didn’t want to turn it on? This will break once we remove rematerialization pass.
> > > > > > 
> > > > > > 
> > > > > > Didnt you plan to move all of that into the softmax decomposition pass?
> > > > > 
> > > > > 
> > > > > Right, the idea is to make the softmax decomposition do the rematerialization when decomposing and then remove the general pass. That means if a code sequence like softmax (but not using the softmax op) is fused we will have large allocations again. It sounds like that’s what could happen with aggressive fusion.
> > > > 
> > > > 
> > > > yup, understood... today rematerialization works, cause really the only case that get fused that way is softmax. Once the decomposition does rematerialzation, I can remove the use of that pass in the backends and make sure we dont fuse in a way that creates allocations. I think those could be decoupled. In any case, large stacks are a hard-error today... so we will know if there are issues.
> > > 
> > > 
> > > Got it, so you are going to remove the cases that would create large allocs from the fusion? That would work. This is not done yet right?
> > 
> > 
> > Trying to guage the order of things. Ill probably defer that to later stage. So keep the rematerialization for now (its used on all backends), and then evaluate after softmax doesnt need it.
> 
> Sounds good to me. I was going to remove it because you were concerned about it but I can’t really think about case where doing the rematerialization is worse and where backend cse wouldn’t revert it in case it is worth. I guess we could have saved the work in matching softmax if we are going to enable aggressive fusion anyway

I dont think it is a waste. Rematerialization and CSE are not the inverse of each other (they only become inverse of each other if you remove all the loops and thats a very special case....). The softmax matcher is needed anyway for flash attention... I am just trying to land it without rocking the boat too much...

**Comment by ThomasRaoux** [comment]
> I dont think it is a waste. Rematerialization and CSE are not the inverse of each other (they only become inverse of each other if you remove all the loops and thats a very special case....).

Right, what I’m saying is that when there are loops we will always end up with stack allocations which is In general not what we want. But let’s if we find counter examples in the future.

**Comment by MaheshRavishankar** [comment]
> 
> > I dont think it is a waste. Rematerialization and CSE are not the inverse of each other (they only become inverse of each other if you remove all the loops and thats a very special case....).
> 
> Right, what I’m saying is that when there are loops we will always end up with stack allocations which is In general not what we want. But let’s if we find counter examples in the future.
> 
> 

Yeah, but we can also not do such fusions. We can say we use rematerialization to avoid stack allocation but that is coming dangerously close to needing a cost model to decide fusion efficacy. That's not the intent right now, and I would like to not rely on it in a way that we cannot turn off materialization if we need based on other examples. Definitely need to look at examples, and today we don't have any. So I am OK with keeping it for now if it isn't used in a load bearing way like it is for softmax.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp [hunk 1]
```diff
@@ -458,12 +412,45 @@ static bool areOpsAggresiveFusable(
       })) {
     return false;
   }
+  return true;
+}
 
-  // Finally only fuse if the `ins` operand can be properly bufferized.
-  // TODO(#10498): Handle the multi-result case.
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchRegions.cpp:464]: What's the status of https://github.com/iree-org/iree/issues/10498?

I think we're able to handle the case for a while. Can you verify it and close the issue?

**Review by hanhanW — APPROVED** [review]
LG!

**Comment by ThomasRaoux** [comment]
> Going to look into the failures, but as it stands these are the dispatches on the benchmarks
> 
> **Before**
> 
> Model	Fills	Copies	Dispatches	Executables	Generic Before	Generic After	DispatchWorkgroups
> bert_encoderdefault_off	11	4	514	41	1198	534	514
> collatzdefault_off	2	1	0	41	12	0	0
> deeplabdefault_off	18	1	78	41	105	57	78
> edge_detectiondefault_off	1	0	3	41	4	1	3
> fragmentdefault_off	3	1	1	41	10	1	1
> fullyconnecteddefault_off	0	0	3	41	8	3	3
> minilmdefault_off	0	2	223	41	1137	249	223
> mnistdefault_off	4	0	6	41	10	6	6
> mobilebert-fp32default_off	25	4	752	41	1741	652	752
> mobilenetv2default_off	18	1	74	41	103	56	74
> mobilessddefault_off	30	1	121	41	157	88	121
> persondetect-int8default_off	14	1	75	41	95	45	75
> posenetdefault_off	14	1	48	41	59	32	48
> resnet50default_off	338	0	78	41	711	55	78
> unidirectional_lstmdefault_off	5	5	10	41	43	7	10
> **After**
> 
> Model	Fills	Copies	Dispatches	Executables	Generic Before	Generic After	DispatchWorkgroups
> bert_encoderdefault_on	11	4	489	41	1198	533	489
> collatzdefault_on	2	1	0	41	12	0	0
> deeplabdefault_on	18	1	78	41	105	57	78
> edge_detectiondefault_on	1	0	3	41	4	1	3
> fragmentdefault_on	3	1	1	41	10	1	1
> fullyconnecteddefault_on	0	0	3	41	8	3	3
> minilmdefault_on	0	2	184	41	1137	248	184
> mnistdefault_on	4	0	4	41	10	6	4
> mobilebert-fp32default_on	1	4	727	41	1741	651	727
> mobilenetv2default_on	18	1	73	41	103	56	73
> mobilessddefault_on	30	1	121	41	157	88	121
> persondetect-int8default_on	14	1	58	41	95	45	58
> posenetdefault_on	14	1	48	41	59	32	48
> resnet50default_on	338	0	76	41	711	55	76
> unidirectional_lstmdefault_on	5	8	8	41	43	5	8

I wonder what those numbers look like after the merge of softmax patterns.

**Comment by MaheshRavishankar** [comment]
> > Going to look into the failures, but as it stands these are the dispatches on the benchmarks
> > **Before**
> > Model	Fills	Copies	Dispatches	Executables	Generic Before	Generic After	DispatchWorkgroups
> > bert_encoderdefault_off	11	4	514	41	1198	534	514
> > collatzdefault_off	2	1	0	41	12	0	0
> > deeplabdefault_off	18	1	78	41	105	57	78
> > edge_detectiondefault_off	1	0	3	41	4	1	3
> > fragmentdefault_off	3	1	1	41	10	1	1
> > fullyconnecteddefault_off	0	0	3	41	8	3	3
> > minilmdefault_off	0	2	223	41	1137	249	223
> > mnistdefault_off	4	0	6	41	10	6	6
> > mobilebert-fp32default_off	25	4	752	41	1741	652	752
> > mobilenetv2default_off	18	1	74	41	103	56	74
> > mobilessddefault_off	30	1	121	41	157	88	121
> > persondetect-int8default_off	14	1	75	41	95	45	75
> > posenetdefault_off	14	1	48	41	59	32	48
> > resnet50default_off	338	0	78	41	711	55	78
> > unidirectional_lstmdefault_off	5	5	10	41	43	7	10
> > **After**
> > Model	Fills	Copies	Dispatches	Executables	Generic Before	Generic After	DispatchWorkgroups
> > bert_encoderdefault_on	11	4	489	41	1198	533	489
> > collatzdefault_on	2	1	0	41	12	0	0
> > deeplabdefault_on	18	1	78	41	105	57	78
> > edge_detectiondefault_on	1	0	3	41	4	1	3
> > fragmentdefault_on	3	1	1	41	10	1	1
> > fullyconnecteddefault_on	0	0	3	41	8	3	3
> > minilmdefault_on	0	2	184	41	1137	248	184
> > mnistdefault_on	4	0	4	41	10	6	4
> > mobilebert-fp32default_on	1	4	727	41	1741	651	727
> > mobilenetv2default_on	18	1	73	41	103	56	73
> > mobilessddefault_on	30	1	121	41	157	88	121
> > persondetect-int8default_on	14	1	58	41	95	45	58
> > posenetdefault_on	14	1	48	41	59	32	48
> > resnet50default_on	338	0	76	41	711	55	76
> > unidirectional_lstmdefault_on	5	8	8	41	43	5	8
> 
> I wonder what those numbers look like after the merge of softmax patterns.

Dont know. Will rebase and rerun scripts.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors_default.mlir [hunk 2]
```diff
@@ -30,38 +30,6 @@ func.func @no_fuse_quantized(%arg0 : tensor<?x113x113x64xi8>, %arg1 : tensor<3x3
...
-// Check that two generic ops are NOT dispatched together since the input type
-// for reduction is different from the output type of the elementwise op. We
-// should see two flow.dispatch.workgroups.
-
-// CHECK-LABEL: func.func @reduction_broadcast_elementwise_type_mismatch
-//      CHECK: flow.dispatch.workgroups
-//      CHECK: flow.dispatch.workgroups
-
-// -----
-
```
Comment (dcaballe) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/test/dispatch_linalg_on_tensors_default.mlir:64]: Do we have tests to cover cases that are now fused and were not before?

**Review by dcaballe — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
Moved the first commit out into its own PR. (#12277 )

**Comment by pzread** [comment]
The compilation errors in CI are related to CUDA backend and ARM CPU backend with data-tiling enabled.

Here is the list of issues:
- https://github.com/iree-org/iree/issues/12287
- https://github.com/iree-org/iree/issues/12286
- https://github.com/iree-org/iree/issues/12285

**Comment by MaheshRavishankar** [comment]
This has landed in main already. Closing

**Pull request closed without merging**